pkg/common/git.go

```diff
@@ -0,0 +1,192 @@
+}
+
+type AbstractGitClient struct {
+	GitClient
+
+	logger    logger.Logger
+	cmdRunner cmdrunner.CmdRunner
+}
+
+func NewGitClient(parentLogger logger.Logger) (GitClient, error) {
+	var err error
+
+	abstractGitClient := AbstractGitClient{logger: parentLogger.GetChild("git-client")}
+
+	// create cmd runner
+	abstractGitClient.cmdRunner, err = cmdrunner.NewShellRunner(parentLogger)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create cmd runner")
+	}
+
+	return abstractGitClient, nil
+}
+
+func (agc AbstractGitClient) Clone(outputDir, repositoryURL string, gitAttributes *GitAttributes) error {
+	var referenceName string
+	var gitAuth *githttp.BasicAuth
+	var err error
+
+	// resolve full git reference name
+	referenceName, err = ResolveGitReference(repositoryURL, gitAttributes)
+	if err != nil {
+		return errors.Wrap(err, "Failed to resolve git reference")
+	}
+
+	// resolve git credentials when given
+	gitAuth = agc.parseFunctionGitCredentials(gitAttributes)
+
+	// if it's Azure Devops repo - clone differently (the normal go-git client doesn't support it yet)
+	if isAzureDevopsRepositoryURL(repositoryURL) {
+		return agc.cloneFromAzureDevops(outputDir, repositoryURL, referenceName, gitAuth, agc.cmdRunner)
+	}
+
+	if _, err = git.PlainClone(outputDir, false, &git.CloneOptions{
+		URL:           repositoryURL,
+		ReferenceName: plumbing.ReferenceName(referenceName),
+		Depth:         1,
+		Auth:          gitAuth,
+	}); err != nil {
+		return errors.Wrap(err, "Failed to clone git repository")
+	}
+
+	agc.printCurrentCommitSHA(outputDir, repositoryURL, referenceName)
+
+	return nil
+}
+
+func (agc AbstractGitClient) printCurrentCommitSHA(gitDir, referenceName, repositoryURL string) {
+	res, err := agc.cmdRunner.Run(nil, fmt.Sprintf("cd %s;git rev-parse HEAD", Quote(gitDir)))
+	if err != nil || res.ExitCode != 0 {
+		agc.logger.WarnWith("Failed to get commit SHA", "err", err)
+		return
+	}
+	if res.ExitCode != 0 {
+		agc.logger.WarnWith("Failed to get commit SHA (non-zero exit code)", "output", res.Output)
+		return
+	}
+
+	// remove automatic new line from end of res.Output
+	commitSHA := strings.TrimSuffix(res.Output, "\n")
+
+	agc.logger.InfoWith("Printing current commit SHA",
+		"repositoryURL", repositoryURL,
+		"referenceName", referenceName,
+		"commitSHA", commitSHA)
+}
+
+func (agc AbstractGitClient) cloneFromAzureDevops(outputDir string,
+	repositoryURL string,
+	referenceName string,
+	gitAuth *githttp.BasicAuth,
+	cmdRunner cmdrunner.CmdRunner) error {
+
+	// if auth is passed, transplant username:password into repository URL
+	if gitAuth != nil {
+		splitFunctionPath := strings.Split(repositoryURL, "://")
+		repositoryURL = fmt.Sprintf("%s://%s:%s@%s",
+			splitFunctionPath[0],
+			gitAuth.Username,
+			gitAuth.Password,
+			splitFunctionPath[1])
+	}
+
+	// generate a git clone command
+	cloneCommand := fmt.Sprintf("git clone %s --depth 1 -q %s",
```

Comment:
git azure devops allowing you to create a Personal access token (for limited access and etc)

the command for that is 

```
git -c http.extraHeader="Authorization: Basic ${B64_PAT}" clone <repo> --branch <branch> --single-branch
```

we can support this by branching  the behavior when user provided `password` only (without username).
B64_PAT is `base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(":%s", password)))`.

you can pass `env` with command runner options struct so that the PAT value would be masked automatically

Reply:
PAT is supported with the current flow..you just need to pass it inside `password` and it works (I tested it)